runtime/src/accounts_partition.rs

```diff
@@ -0,0 +1,718 @@
+        );
+        assert_eq!(
+            end_index,
+            partition_from_pubkey(&end_pubkey_final, partition_count),
+            "{start_index}, {end_index}, {end_pubkey_final}, {partition_count}"
+        );
+        if start_index != 0 {
+            start_pubkey[0..PREFIX_SIZE]
+                .copy_from_slice(&start_key_prefix.saturating_sub(1).to_be_bytes());
+            let pubkey_test = Pubkey::new_from_array(start_pubkey);
+            assert_eq!(
+                start_index,
+                partition_from_pubkey(&pubkey_test, partition_count),
+                "{}, {}, start_key_prefix-1: {}, {}, {}",
+                start_index,
+                end_index,
+                start_key_prefix.saturating_sub(1),
+                pubkey_test,
+                partition_count
+            );
+        }
+        if end_index != partition_count - 1 && end_index != 0 {
+            end_pubkey[0..PREFIX_SIZE]
+                .copy_from_slice(&end_key_prefix.saturating_add(1).to_be_bytes());
+            let pubkey_test = Pubkey::new_from_array(end_pubkey);
+            assert_eq!(
+                end_index.saturating_add(1),
+                partition_from_pubkey(&pubkey_test, partition_count),
+                "start: {}, end: {}, pubkey: {}, partition_count: {}, prefix_before_addition: {}, prefix after: {}",
+                start_index,
+                end_index,
+                pubkey_test,
+                partition_count,
+                end_key_prefix,
+                end_key_prefix.saturating_add(1),
+            );
+        }
+    }
+    // should be an inclusive range (a closed interval) like this:
+    // [0xgg00-0xhhff], [0xii00-0xjjff], ... (where 0xii00 == 0xhhff + 1)
+    start_pubkey_final..=end_pubkey_final
+}
+
+pub(crate) fn prefix_from_pubkey(pubkey: &Pubkey) -> u64 {
+    const PREFIX_SIZE: usize = mem::size_of::<u64>();
+    u64::from_be_bytes(pubkey.as_ref()[0..PREFIX_SIZE].try_into().unwrap())
+}
+
+/// This is the inverse of pubkey_range_from_partition.
+/// return the lowest end_index which would contain this pubkey
+pub(crate) fn partition_from_pubkey(
+    pubkey: &Pubkey,
+    partition_count: PartitionsPerCycle,
+) -> PartitionIndex {
+    type Prefix = u64;
+    const PREFIX_MAX: Prefix = Prefix::max_value();
+
+    if partition_count == 1 {
+        return 0;
+    }
+
+    // not-overflowing way of `(Prefix::max_value() + 1) / partition_count`
+    let partition_width = (PREFIX_MAX - partition_count + 1) / partition_count + 1;
+
+    let prefix = prefix_from_pubkey(pubkey);
+    if prefix == 0 {
+        return 0;
+    }
+
+    if prefix == PREFIX_MAX {
+        return partition_count - 1;
+    }
+
+    let mut result = (prefix + 1) / partition_width;
+    if (prefix + 1) % partition_width == 0 {
+        // adjust for integer divide
+        result = result.saturating_sub(1);
+    }
+    result
+}
+
+lazy_static! {
+    pub static ref EMPTY_HASHSET: HashSet<Pubkey> = HashSet::default();
```

Comment:
I think it makes sense to adjust visibilities with this move.

This is not used outside this file. Doesn't need to be public.

Below: `RentPayingAccountsByPartition` is only used in runtime, so we can make it pub(crate).

Those were the only pub facing things in the module, so let's also make the module in lib non-pub 😄

Reply:
That's true, but `RentPayingAccountsByPartition` is used in pub structures such as `IndexGenerationInfo`, so compiler will complain about leaking crate private type :(